### ISSUE TITLE
isConstructType excludes every subclass of Stack, disabling three rules and misreporting Stack-typed scopes

### ISSUE THREAD

**Issue by ren-yamanashi** [ren-yamanashi/awscdk-lint#570]
### Describe the bug

`isConstructType()` takes an `ignoredClasses` list that defaults to
`["App", "Stage", "CfnOutput", "Stack"]`, and `isExtendsFromTargetSuperClass()` recurses through
`isConstructType` itself when walking base types. As a result the exclusion is not limited to
the `Stack` class — it propagates to the whole `Stack` subtree, so **any** `class X extends Stack`
is treated as "not a Construct".

Three rules gate on `isConstructType()` and therefore do nothing at all for stack classes:

- `no-unused-props`
- `construct-constructor-property`
- `require-jsdoc` (the Construct public-property half)

Neither the rule docs nor the tests mention a `Stack` exclusion for these rules
(`docs/rules/no-unused-props.md`, `docs/rules/construct-constructor-property.md`,
`docs/rules/require-jsdoc.md`). `props-name-convention` is a deliberate exception — its doc
does say "This rule applies only to classes that extends from `Construct`".

The same leak also produces a **false positive**: `construct-constructor-property` checks the
`scope` parameter's type with `isConstructType()`, so a `Stack`-typed scope is rejected even
though `Stack` is a `Construct`.

`class MyStack extends Stack` with a `MyStackProps` interface is one of the most common CDK
shapes, so this is a large blind spot.

**Reproduction 1 — `no-unused-props`**

`src/fixture.ts`:

```ts
import { Stack } from "aws-cdk-lib";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface MyProps {
  readonly usedProp: string;
  readonly unusedProp: string;
}

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: MyProps) {
    super(scope, id);
    new Bucket(this, "Bucket", { bucketName: props.usedProp });
  }
}

export class MyConstruct extends Construct {
  constructor(scope: Construct, id: string, props: MyProps) {
    super(scope, id);
    new Bucket(this, "Bucket", { bucketName: props.usedProp });
  }
}
```

```
$ npx eslint --config eslint.config.mjs src/fixture.ts

/path/to/repro/src/fixture.ts
  18:45  error  Property 'unusedProp' is defined in props but never used  awscdk/no-unused-props

✖ 1 problem (1 error, 0 warnings)

$ npx oxlint --format=unix -c .oxlintrc.json src/fixture.ts
src/fixture.ts:18:45: Property 'unusedProp' is defined in props but never used [Error/awscdk(no-unused-props)]

1 problem
```

Only `MyConstruct` (line 18) is reported; the identical `MyStack` (line 11) is not.

**Reproduction 2 — `construct-constructor-property`**

`src/fixture.ts`:

```ts
import { Stack } from "aws-cdk-lib";
import { Construct } from "constructs";

export class BadStack extends Stack {
  constructor(myScope: Construct, myId: string) {
    super(myScope, myId);
  }
}

export class BadConstruct extends Construct {
  constructor(myScope: Construct, myId: string) {
    super(myScope, myId);
  }
}
```

```
$ npx eslint --config eslint.config.mjs src/fixture.ts

/path/to/repro/src/fixture.ts
  11:15  error  Constructor of a class extending Construct must have the property names 'scope, id' or 'scope, id, props'  awscdk/construct-constructor-property
  11:35  error  Constructor of a class extending Construct must have the property names 'scope, id' or 'scope, id, props'  awscdk/construct-constructor-property

✖ 2 problems (2 errors, 0 warnings)

$ npx oxlint --format=unix -c .oxlintrc.json src/fixture.ts
src/fixture.ts:11:15: Constructor of a class extending Construct must have the property names 'scope, id' or 'scope, id, props' [Error/awscdk(construct-constructor-property)]
src/fixture.ts:11:35: Constructor of a class extending Construct must have the property names 'scope, id' or 'scope, id, props' [Error/awscdk(construct-constructor-property)]

2 problems
```

**Reproduction 3 — `require-jsdoc`**

`src/fixture.ts`:

```ts
import { Stack } from "aws-cdk-lib";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export class MyStack extends Stack {
  public readonly bucket!: IBucket;
}

export class MyConstruct extends Construct {
  public readonly bucket!: IBucket;
}
```

```
$ npx eslint --config eslint.config.mjs src/fixture.ts

/path/to/repro/src/fixture.ts
  10:3  error  Property 'bucket' should have a JSDoc comment  awscdk/require-jsdoc

✖ 1 problem (1 error, 0 warnings)

$ npx oxlint --format=unix -c .oxlintrc.json src/fixture.ts
src/fixture.ts:10:3: Property 'bucket' should have a JSDoc comment. [Error/awscdk(require-jsdoc)]

1 problem
```

**Reproduction 4 — `construct-constructor-property`, false positive on a `Stack`-typed scope**

`src/fixture.ts`:

```ts
import { Stack } from "aws-cdk-lib";
import { Construct } from "constructs";

export class MyStack extends Stack {}

export class ScopeIsConstruct extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);
  }
}

export class ScopeIsStack extends Construct {
  constructor(scope: Stack, id: string) {
    super(scope, id);
  }
}

export class ScopeIsStackSubclass extends Construct {
  constructor(scope: MyStack, id: string) {
    super(scope, id);
  }
}
```

```
$ npx eslint --config eslint.config.mjs src/fixture.ts

/path/to/repro/src/fixture.ts
  13:15  error  Constructor of a class extending Construct must have the type 'Construct' for the first parameter  awscdk/construct-constructor-property
  19:15  error  Constructor of a class extending Construct must have the type 'Construct' for the first parameter  awscdk/construct-constructor-property

✖ 2 problems (2 errors, 0 warnings)

$ npx oxlint --format=unix -c .oxlintrc.json src/fixture.ts
src/fixture.ts:13:15: Constructor of a class extending Construct must have the type 'Construct' for the first parameter [Error/awscdk(construct-constructor-property)]
src/fixture.ts:19:15: Constructor of a class extending Construct must have the type 'Construct' for the first parameter [Error/awscdk(construct-constructor-property)]

2 problems
```

Narrowing a construct's scope to `Stack` (or to a project-specific stack base class) is ordinary
CDK, and `super(scope, id)` type-checks fine.

**Configs used** (rule name swapped per reproduction)

`eslint.config.mjs`:

```js
import cdkPlugin from "eslint-plugin-awscdk";
import tseslint from "typescript-eslint";

export default [
  {
    files: ["**/*.ts"],
    plugins: { awscdk: cdkPlugin },
    languageOptions: {
      parser: tseslint.parser,
      parserOptions: { projectService: true, tsconfigRootDir: import.meta.dirname },
    },
    rules: { "awscdk/no-unused-props": "error" },
  },
];
```

`.oxlintrc.json`:

```json
{
  "jsPlugins": ["oxlint-plugin-awscdk"],
  "options": { "typeAware": true },
  "rules": { "awscdk/no-unused-props": "error" }
}
```

### Expected Behavior

`no-unused-props`, `construct-constructor-property` and `require-jsdoc` should apply to classes
that extend `Stack`, since `Stack` is itself a `Construct` and the rule documentation draws no
distinction. `constructor(scope: Stack, id: string)` should not be reported as having a
non-`Construct` first parameter. (`no-mutable-public-property-of-construct` and
`no-construct-in-public-property-of-construct`, which use `isConstructOrStackType()`, already
behave this way.)

### Current Behavior

Every `class X extends Stack` (and, by the same mechanism, `extends App` / `extends Stage`)
is treated as not-a-Construct: skipped by the three rules above, and rejected when it appears as
the `scope` parameter type.

Confirmed with the TypeScript compiler API using the plugin's own logic:

```
MyStack:      symbol=MyStack      isConstructType=false bases=[Stack]
MyConstruct:  symbol=MyConstruct  isConstructType=true  bases=[Construct]
BadStack:     symbol=BadStack     isConstructType=false bases=[Stack]
BadConstruct: symbol=BadConstruct isConstructType=true  bases=[Construct]
```

### Plugin

eslint-plugin-awscdk, oxlint-plugin-awscdk

### Plugin Version

eslint-plugin-awscdk 4.4.1, oxlint-plugin-awscdk 1.0.1

### Language Version

TypeScript 6.0.3 (eslint-plugin-awscdk) / TypeScript 7.0.2 via oxlint-tsgolint (oxlint-plugin-awscdk)

### Additional Context

```yaml
# machine-readable summary
rules: ["no-unused-props", "construct-constructor-property", "require-jsdoc"]
symptom: ["false-negative", "false-positive"]
affected_packages: ["eslint-plugin-awscdk", "oxlint-plugin-awscdk"]
root_cause: "packages/*/src/core/ts-type/checker/is-extends-target-super-class.ts:9"
verified_on: "main @ 6044b32 (eslint-plugin-awscdk 4.4.1, oxlint-plugin-awscdk 1.0.1)"
related_issues: ["#54", "#78", "#93", "#428"]
```

Root cause: the ignore list is consulted at every level of the base-type walk, not just for the
type under test.

- `packages/eslint/src/core/cdk-construct/type-checker/is-construct.ts:11`
- `packages/eslint/src/core/ts-type/checker/is-extends-target-super-class.ts:9`
- `packages/oxlint/src/core/cdk-construct/type-checker/is-construct.ts:12`
- `packages/oxlint/src/core/ts-type/checker/is-extends-target-super-class.ts:11`

Applying `ignoredClasses` only to the *initial* type (and not while recursing into base types)
would keep `new Stack(...)` itself out of scope while restoring coverage for stack subclasses.

**Call sites of `isConstructType` (identical in both packages), grouped by what is passed in**

| Argument | Call sites | Effect of the leak |
|---|---|---|
| type of a `new` expression | `no-parent-name-construct-id-match:59`, `prevent-construct-id-collision:33`, `require-passing-this:58`, `no-variable-construct-id:36` | the ignore list is *wanted* here (see #78 / #93) |
| type of a `ClassDeclaration` | `construct-constructor-property:54`, `no-unused-props:53`, `require-jsdoc:76` | unintended — reproductions 1–3 |
| type of the `scope` parameter binding | `construct-constructor-property:102` | unintended — reproduction 4 |
| type of `superClass` | `props-name-convention:34` | intended (doc says Construct-only) |
| receiver of a `grant*` call | `prefer-grants-property:39` | unrelated to this issue |

**Where the ignore list came from, and what to watch when changing it**

- **#78** (`new SampleStack(app, 'Sample')` must not trigger `require-passing-this`) and **#93**
  (`new App({})` must not either) are why `Stack` / `App` are on the list; **#54** asked for
  rules to apply only to Construct/Stack classes. That intent is correct for the
  new-expression call sites only.
- Applying `ignoredClasses` only to the *initial* type (not while recursing into base types) is
  the minimal fix, but it re-arms the four new-expression call sites for `class X extends Stack`.
  The #78 and #93 snippets are additionally protected today by the `findEnclosingClass()` guard
  added for #427 / #432 — verified: both snippets report nothing on `main`, and the guard alone
  covers top-level instantiation. What would newly become visible is instantiating a stack
  subclass *inside* a Construct/Stack class; decide deliberately whether that is wanted.
- `props-name-convention` is documented as Construct-only. With a top-level-only ignore it stays
  correct for `class X extends Stack` (the `superClass` symbol is literally `Stack`) but would
  **start** applying to `class X extends MyBaseStack` where `MyBaseStack extends Stack`.
- **#428** (open) is the mirror-image complaint — `isConstructType` is *too narrow* when the base
  class comes from a third-party library. Both point at
  `core/ts-type/checker/is-extends-target-super-class.ts`; fixing them together avoids two rounds
  of churn in the same helper.

There are no unit tests covering `extends Stack` for `no-unused-props`,
`construct-constructor-property` or `require-jsdoc`, which is consistent with this being an
unintended gap rather than a documented decision.

Reproduction environment:

- macOS (darwin 23.1.0), Node.js v24.18.0, pnpm 10.29.3
- aws-cdk-lib 2.266.0 / constructs 10.8.1
- ESLint 10.8.1 + typescript-eslint 8.67.0 + TypeScript 6.0.3
- oxlint 1.79.0 + oxlint-tsgolint 7.0.2001 (corsa-oxlint 1.12.2)
